### ISSUE TITLE
Consider overloadding `==` and `!=` for `IPAddress`

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#70978]
We hit this in https://github.com/dotnet/runtime/pull/57250 and right know hitting here:
https://github.com/dotnet/runtime/blob/38794de5f2e3198b28df0915e82e33ab3d91f2ee/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs#L191

Customer hitting: https://github.com/dotnet/runtime/issues/38040

However, this is highly likely breaking change (anyone expecting `==` to compare references), so at least discuss this.

**Comment by skyoxZ** [comment]
I'm curious why `IPAddress.Equals(IPAddress other)` is internal.

**Comment by karelz** [comment]
Triage: We should take it to make devs lives better. Especially given that even we ourselves made the mistake at least 2x already.
It will be source breaking change (we need to document it) -- let's try to do it early in 8.0 (a bit too risky in 7.0).

**Comment by danmoseley** [comment]
@bartonjs  what do the design guidelines have to say about the behavior of == differing from the behavior of .Equals()?

**Comment by bartonjs** [comment]
The guideline is

> DO ensure that Object.Equals and the equality operators have exactly the same semantics and similar performance characteristics.

(Framework Design Guidelines, 3rd edition, section 8.13 (Equality Operators))

**Comment by danmoseley** [comment]
@stephentoub have we ever experimented with an analyzer that detects when `==` is overloaded and `Equals(object?)` is not overridden, or vice versa?

We probably have many cases of this, even in public classes. Random example: MailAddress.

**Comment by stephentoub** [comment]
> have we ever experimented with an analyzer that detects when == is overloaded and Equals(object?) is not overridden, or vice versa?

There's:
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1815
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/CA2224
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/CA2231

We've not enabled them because they're really noisy given our current state of things.

**Comment by danmoseley** [comment]
I wonder whether it would make sense to enable and baseline warnings (if that's possible) so we don't add more.

**Comment by wfurt** [comment]
should we even drag it forward @karelz? I feel that the risk is high and the workaround to disable  the rule is simple for anybody who needs the perf.

**Comment by karelz** [comment]
@wfurt we saw some value - https://github.com/dotnet/runtime/issues/70978#issuecomment-1162003957. It is IMO worth a discussion. Not now though.

**Comment by wfurt** [comment]
triage: not critical for 9.0, moving to future

**Comment by julealgon** [comment]
> triage: not critical for 9.0, moving to future

Will this _ever_ be "critical" for any .NET release though? This is the wrong way to approach/classify these enhancements IMHO.

**Comment by skyoxZ** [comment]
> > triage: not critical for 9.0, moving to future
> 
> Will this _ever_ be "critical" for any .NET release though? This is the wrong way to approach/classify these enhancements IMHO.

It's not only an enhancement but also a breaking change, so I guess it won't be fixed unless enough reasons collected.

**Comment by wfurt** [comment]
> > > triage: not critical for 9.0, moving to future
> > 
> > 
> > Will this _ever_ be "critical" for any .NET release though? This is the wrong way to approach/classify these enhancements IMHO.
> 
> It's not only an enhancement but also a breaking change, so I guess it won't be fixed unless enough reasons collected.

right. While it feels useful I'm not sure it is worth of the troubles. I'm not sure if it os worth to keep it open @manickap

**Comment by JamesNK** [comment]
An analyzer wouldn't be breaking. e.g. look for code that combines `==` or `!=` with `IPAddress` on both sides of the operator. If you want to be more specific, require `IPAddress.IPv6Any` or `IPAddress.Any` on one side of the comparison. The analyzer would suggest using `Equals` instead.

It's caused at least two bugs in dotnet/runtime, one customer reported bug. It's not _critical_, but it can be improved.

**Comment by ManickaP** [comment]
I think we should do it, implement the equality. Quick look at grepp.app: https://grep.app/search?q=%3D%3D%20IPAddress%5C.&regexp=true&filter[lang][0]=C%23 and the few random places I checked look like they at least wouldn't logically break with the change. 
Honestly, let's get it over with and get it done now. If we have someone really screaming, we can always introduce a switch to bring back the old behavior, but I doubt that anyone will need it.

@dotnet/ncl thoughts?

**Comment by rzikm** [comment]
Fine with me.

The change seems simple enough, I think this should be also easy for external contributors to pick up if we greenlight it.

**Comment by antonfirsov** [comment]
I'm rather in the analyzer camp, but not a strong opinion. However, if we do the breaking change, it's better in 9 than in an LTS.

**Comment by wfurt** [comment]
I'm fine with proceeding if you want to champion it @ManickaP. From triage prospective I still feel it can stay in future e.g. it is nice to have but it would not  block 9.0 release.

**Comment by ero-qt** [comment]
@ManickaP @wfurt What's the status here? Is this free to take on?

**Comment by ManickaP** [comment]
@just-ero we are OK with implementing this. We'll need to go through API review process for this, only after then we can merge any implementation. 
I think it makes sense to extract it and have it as a separate API, as you did in https://github.com/dotnet/runtime/issues/119952. So feel free to re-open and if you wish put up a draft PR for it. I'll put it on API review radar.
